From 5d38a20196b9b9d15552ba3617ec7f63f6401a5c Mon Sep 17 00:00:00 2001
From: Ard Biesheuvel <ardb@kernel.org>
Date: Tue, 25 Nov 2025 14:29:20 +0100
Subject: [PATCH] BaseTools/tools_def: Use LLD to link ACPI table binaries

Currently, the CLANGDWARF toolchain on AARCH64 fails to pass the
-fuse-ld=lld compiler option, and so linking ACPI table binaries falls
back to the BFD linker. On non-AARCH64 build systems, this will
correctly run the cross-linker, based on the -target argument passed to
Clang, which therefore needs to be installed.

However, the ASLCC invocation fails to pass that same -target argument,
therefore producing object files for the native architecture, which the
cross-linker cannot link, resulting in a build error.

So pass the target in ASLCC_FLAGS as well, which is sufficient to get a
working build. And for good measure, pass -fuse-ld=lld so that we don't
rely on the BFD cross-linker in the first place.

Signed-off-by: Ard Biesheuvel <ardb@kernel.org>

Upstream-Status: Backport [5d38a20196b9b9d15552ba3617ec7f63f6401a5c]
Signed-off-by: Jon Mason <jon.mason@arm.com>
---
 BaseTools/Conf/tools_def.template | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/BaseTools/Conf/tools_def.template b/BaseTools/Conf/tools_def.template
index 32085ad00af7..d19624fa4628 100755
--- a/BaseTools/Conf/tools_def.template
+++ b/BaseTools/Conf/tools_def.template
@@ -1872,8 +1872,8 @@ DEFINE CLANGDWARF_AARCH64_DLINK_FLAGS  = DEF(CLANGDWARF_AARCH64_TARGET) DEF(GCC_
 *_CLANGDWARF_AARCH64_SLINK_PATH     = ENV(CLANGDWARF_BIN)llvm-ar
 *_CLANGDWARF_AARCH64_RC_PATH        = ENV(CLANGDWARF_BIN)llvm-objcopy
 
-*_CLANGDWARF_AARCH64_ASLCC_FLAGS    = DEF(GCC_ASLCC_FLAGS) -fno-lto
-*_CLANGDWARF_AARCH64_ASLDLINK_FLAGS = DEF(CLANGDWARF_AARCH64_TARGET) DEF(GCC_AARCH64_ASLDLINK_FLAGS) DEF(CLANGDWARF_DLINK_WARNING_FLAGS)
+*_CLANGDWARF_AARCH64_ASLCC_FLAGS    = DEF(GCC_ASLCC_FLAGS) DEF(CLANGDWARF_AARCH64_TARGET) -fno-lto
+*_CLANGDWARF_AARCH64_ASLDLINK_FLAGS = DEF(CLANGDWARF_AARCH64_TARGET) DEF(GCC_AARCH64_ASLDLINK_FLAGS) DEF(CLANGDWARF_DLINK_WARNING_FLAGS) -fuse-ld=lld
 *_CLANGDWARF_AARCH64_ASM_FLAGS      = DEF(GCC_ASM_FLAGS) DEF(CLANGDWARF_AARCH64_TARGET) $(PLATFORM_FLAGS) -Qunused-arguments
 *_CLANGDWARF_AARCH64_DLINK_FLAGS    = DEF(CLANGDWARF_AARCH64_TARGET) DEF(GCC_AARCH64_DLINK_FLAGS) -z common-page-size=0x1000 DEF(CLANGDWARF_DLINK_WARNING_FLAGS)
 *_CLANGDWARF_AARCH64_DLINK_XIPFLAGS = -z common-page-size=0x20
